### PR TITLE
RISC-V: Implement stack unwinding

### PR DESCRIPTION
**Pull request by gagachang**
This patch sets implement the kernel stack unwinding for RISC-V architecture.
The detail can be found in the commit message of [libunw: Implement RISC-V stack unwinding](https://github.com/OP-TEE/optee_os/commit/904034a75ccf31c59831f23cf704dd6012aa5eb6)

### CODE REVIEW

**Comment by gagachang** [comment]
Dear @liushiwei007 @maroueneboubakri 

Would you have time to review this PR?
Any suggestions are appreciated!

**Comment by jforissier** [comment]
Hi @gagachang,

I believe that `scripts/symbolize.py` needs a few adaptations for RISC-V. Could you please take a look? It would be nice to have `symbolize.py` be part of this unwinding PR, because obviously stack dumps are immensely more helpful when they can be decoded. Also, the script is a good way to validate that the unwinding is correct.
Thanks!

**Comment by gagachang** [comment]
> Hi @gagachang,
> 
> 
> 
> I believe that `scripts/symbolize.py` needs a few adaptations for RISC-V. Could you please take a look? It would be nice to have `symbolize.py` be part of this unwinding PR, because obviously stack dumps are immensely more helpful when they can be decoded. Also, the script is a good way to validate that the unwinding is correct.
> 
> Thanks!

Sure!
I will check it and push one more commit.

**Comment by gagachang** [comment]
Hi @jforissier 

I found a problem in the script and need some recommendations:

In `symbol_plus_offset()`, the **addr** is stripped first because of ARM MTE:
https://github.com/OP-TEE/optee_os/blob/master/scripts/symbolize.py#L261

RISC-V can skip this by checking `self._arch`, but the `self._arch` is parsed latter at
https://github.com/OP-TEE/optee_os/blob/master/scripts/symbolize.py#L267
because the ELF is got from **addr**.

Should we add another command-line argument to solve the problem?
Such like `./optee_os/scripts/symbolize.py -a {arm|risc-v}`.

**Comment by jforissier** [comment]
Hello @gagachang,

> Should we add another command-line argument to solve the problem?
> Such like `./optee_os/scripts/symbolize.py -i {arm|risc-v}`.
> And we can get another argument called `self._isa` in this case.

I'd like to avoid that if possible. I propose removing the tag in the stack dump instead so that `symbolize.py` does not need to do it. Please see https://github.com/OP-TEE/optee_os/pull/6062.

**Comment by gagachang** [comment]
Hello @jforissier 

I've added a new commit for `symbolize.py`, please check it.
Thanks!

**Comment by jforissier** [comment]
> Hello @jforissier
> 
> I've added a new commit for `symbolize.py`, please check it. Thanks!

Please also remove `(arm-linux-gnueabihf-, aarch64-linux-gnu-)` from the help text. With that, for the commit:

`Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`

**Comment by gagachang** [comment]
> Please also remove `(arm-linux-gnueabihf-, aarch64-linux-gnu-)` from the help text. With that, for the commit:
> 
> `Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`

Done!
Is the acknowledgement also applicable to other commits?

**Comment by jforissier** [comment]
> Is the acknowledgement also applicable to other commits?

No :wink:

**Comment by gagachang** [comment]
> > Is the acknowledgement also applicable to other commits?
> 
> No 😉

OK~ I will wait for it 😆

**Comment by liushiwei007** [comment]
> Dear @liushiwei007 @maroueneboubakri
> 
> Would you have time to review this PR? Any suggestions are appreciated!

hi, gagachang 
Sorry, I haven't actually done this part. I can follow along and learn this part of your code.

**Comment by gagachang** [comment]
> > Dear @liushiwei007 @maroueneboubakri
> > Would you have time to review this PR? Any suggestions are appreciated!
> 
> hi, gagachang Sorry, I haven't actually done this part. I can follow along and learn this part of your code.

OK!
If there are any issues please let me know.

**Comment by jforissier** [comment]
@gagachang have you tested the crash dump with the `symbolize.py script`? I'd like to see the output. For example forcing a crash with this patch:
```diff
diff --git a/core/mm/tee_mm.c b/core/mm/tee_mm.c
index c61306d75..94e371889 100644
--- a/core/mm/tee_mm.c
+++ b/core/mm/tee_mm.c
@@ -65,6 +65,7 @@ bool tee_mm_init(tee_mm_pool_t *pool, paddr_t lo, paddr_size_t size,
        pool->entry->pool = pool;
        pool->lock = SPINLOCK_UNLOCK;

+       *(char *)0 = 1;
        return true;
 }

```
Here is the output on Arm (AArch64):
```
$ cat /tmp/crash.txt | ./optee_os/scripts/symbolize.py -d optee_os/out/arm/core
E/TC:0 0
E/TC:0 0 Core data-abort at address 0x0 (translation fault)
E/TC:0 0  esr 0x96000046  ttbr0 0x0e1b1000   ttbr1 0x00000000   cidr 0x0
E/TC:0 0  cpu #0          cpsr 0x800003c4
E/TC:0 0  x0  0000000000000000 x1  0000000000000000
E/TC:0 0  x2  0000000039ce6ff0 x3  0000008000000000
E/TC:0 0  x4  0000000000000070 x5  0000000039d04f50
E/TC:0 0  x6  ffffffffffffffd0 x7  0000000000010bc0
E/TC:0 0  x8  0000000039d04fa0 x9  0000000039cf4360
E/TC:0 0  x10 0000000039cc8118 x11 0000000000000000
E/TC:0 0  x12 0000000039d19064 x13 0000000039d18ebd
E/TC:0 0  x14 0000000039d190c0 x15 0000000039cc8118
E/TC:0 0  x16 0000000039c741c8 x17 0000000000000000
E/TC:0 0  x18 0000000039d19064 x19 0000000000d00000
E/TC:0 0  x20 0000000039ce85f0 x21 0000000000000000
E/TC:0 0  x22 000000000000000c x23 0000000000000000
E/TC:0 0  x24 0000000000000000 x25 0000000000000000
E/TC:0 0  x26 0000000000000000 x27 0000000000000000
E/TC:0 0  x28 0000000000000000 x29 0000000039d19190
E/TC:0 0  x30 0000000039c84f20 elr 0000000039c84f48
E/TC:0 0  sp_el0 0000000039d19190
E/TC:0 0 TEE load address @ 0x39c63000
E/TC:0 0 Call stack:
E/TC:0 0  0x39c84f48 tee_mm_init at optee_os/core/mm/tee_mm.c:66
E/TC:0 0  0x39c6ad60 init_tee_runtime at optee_os/core/arch/arm/kernel/boot.c:1368
E/TC:0 0  0x39c6b238 boot_init_primary_late at optee_os/core/arch/arm/kernel/boot.c:1479
E/TC:0 0 Panic 'unhandled pageable abort' at core/arch/arm/kernel/abort.c:580 <abort_handler>
E/TC:0 0 TEE load address @ 0x39c63000
E/TC:0 0 Call stack:
E/TC:0 0  0x39c6b5a4 print_kernel_stack at optee_os/core/arch/arm/kernel/unwind_arm64.c:89
E/TC:0 0  0x39c78c68 __do_panic at optee_os/core/kernel/panic.c:24
E/TC:0 0  0x39c6a070 abort_handler at optee_os/core/arch/arm/kernel/abort.c:580
E/TC:0 0  0x39c673b0 el1_sync_abort at optee_os/core/arch/arm/kernel/thread_a64.S:964
```
```

**Comment by gagachang** [comment]
Hi, @jforissier 

I got the following information from `symbolize.py`.

```
cat dump.txt | ./optee_os/scripts/symbolize.py -d ./optee_os/out/riscv/core/
E/TC:0 0 Core data-abort at address 0x0 (translation fault)
E/TC:0 0 cpu    #0
E/TC:0 0 cause  000000000000000f epc    000000008e023786
E/TC:0 0 tval   0000000000000000 status 8000000200046100
E/TC:0 0 ra     000000008e02372e sp     000000008e0ce690
E/TC:0 0 gp     000000008e086268 tp     0000000000000018
E/TC:0 0 t0     0000000000000001 t1     0000000000000800
E/TC:0 0 t2     000000008e000156 s0     000000008e0ce790
E/TC:0 0 s1     000000008e0ce6b0 a0     000000008e0b2fb0
E/TC:0 0 a1     0000000000000222 a2     0000000000000018
E/TC:0 0 a3     0000000000000000 a4     0000000000000000
E/TC:0 0 a5     0000000000000001 a5     0000000000000001
E/TC:0 0 a6     0000000091c00000 a7     0000000091dfffff
E/TC:0 0 s2     000000008e05b83e s3     000002228e0b2fb0
E/TC:0 0 s4     000000008e086268 s5     000000008e0ce720
E/TC:0 0 s6     000000008e05be54 s7     0000000000000018
E/TC:0 0 s8     0000000000000001 s9     0000022200d00000
E/TC:0 0 s10    000000008e0b2fb0 s11    000000008e0ce750
E/TC:0 0 t3     000000008001902e t4     000000008004ddaf
E/TC:0 0 t5     0000000000000001 t6     fffffffffffffff6
E/TC:0 0 TEE load address @ 0x8e000000
E/TC:0 0 Call stack:
E/TC:0 0  0x8e023786 tee_mm_init at optee_os/core/mm/tee_mm.c:62
E/TC:0 0 Panic 'unhandled page fault abort' at ?:0
E/TC:0 0 TEE load address @ 0x8e000000
E/TC:0 0 Call stack:
E/TC:0 0  0x8e0046aa read_fp at optee_os/core/arch/riscv/kernel/unwind_rv.c:143
E/TC:0 0  0x8e004768 print_kernel_stack at optee_os/core/arch/riscv/kernel/unwind_rv.c:51
E/TC:0 0  0x8e013a92 __do_panic at optee_os/core/kernel/panic.c:26 (discriminator 28)
E/TC:0 0  0x8e001690 abort_handler at optee_os/core/arch/riscv/kernel/abort.c:384
E/TC:0 0  0x8e001f6e thread_abort_handler at optee_os/core/arch/riscv/kernel/thread_arch.c:223
E/TC:0 0  0x8e001fc0 thread_exception_handler at optee_os/core/arch/riscv/kernel/thread_arch.c:231
E/TC:0 0  0x8e00208e thread_trap_handler at optee_os/core/arch/riscv/kernel/thread_arch.c:278
E/TC:0 0  0x8e0002a0 trap_from_kernel at optee_os/core/arch/riscv/kernel/thread_rv.S:124
E/TC:0 0  0x8e02347e core_mmu_init_ta_ram at optee_os/core/mm/core_mmu.c:2520
E/TC:0 0  0x8e000938 init_tee_runtime at optee_os/core/arch/riscv/kernel/boot.c:61
E/TC:0 0  0x8e000a66 boot_init_primary_late at optee_os/core/arch/riscv/kernel/boot.c:119
E/TC:0 0  0x8e0000d4 reset_primary at optee_os/core/arch/riscv/kernel/entry.S:173
```

**Comment by gagachang** [comment]
I am not sure whether the first call stack is correct or not.

```
E/TC:0 0 Call stack:
E/TC:0 0  0x8e023786 tee_mm_init at optee_os/core/mm/tee_mm.c:62
```

If it is problem I will check it.

**Comment by gagachang** [comment]
Hmm, I guess the problem is stack pointer.
In RISC-V's entry.S, the stack pointer is not assigned as thread stack between `boot_init_primary_early()` and `boot_init_primary_late()`.

Will check it later.

**Comment by jforissier** [comment]
> I am not sure whether the first call stack is correct or not. It's too short.
> 
> ```
> E/TC:0 0 Call stack:
> E/TC:0 0  0x8e023786 tee_mm_init at optee_os/core/mm/tee_mm.c:62
> ```

Indeed it should unwind down to `reset_primary()` I suppose.  The panic stack looks absolutely fine though. Very good for debugging :wink:

**Comment by gagachang** [comment]
Hi @jforissier 

> Hmm, I guess the problem is stack pointer.
In RISC-V's entry.S, the stack pointer is not assigned as thread stack between boot_init_primary_early() and boot_init_primary_late().

After fixing the stack pointer between those two functions, I got correct unwinding.
I think this PR is good.
I will create another PR to fix that after more checking.

```
cat dump.txt | ./optee_os/scripts/symbolize.py -d ./optee_os/out/riscv/core/
E/TC:0 0 Core data-abort at address 0x0 (translation fault)
E/TC:0 0 cpu    #0
E/TC:0 0 cause  000000000000000f epc    000000008e0237a8
E/TC:0 0 tval   0000000000000000 status 8000000200046100
E/TC:0 0 ra     000000008e023750 sp     000000008e0d1290
E/TC:0 0 gp     000000008e086268 tp     0000000000000018
E/TC:0 0 t0     0000000000000001 t1     000000008e0d1400
E/TC:0 0 t2     000000008e00016e s0     000000008e0d1390
E/TC:0 0 s1     000000008e0d12b0 a0     000000008e0b2fb0
E/TC:0 0 a1     0000000000000222 a2     0000000000000018
E/TC:0 0 a3     0000000000000000 a4     0000000000000000
E/TC:0 0 a5     0000000000000001 a5     0000000000000001
E/TC:0 0 a6     0000000091c00000 a7     0000000091dfffff
E/TC:0 0 s2     000000008e05b860 s3     000002228e0b2fb0
E/TC:0 0 s4     000000008e086268 s5     000000008e0d1320
E/TC:0 0 s6     000000008e05be76 s7     0000000000000018
E/TC:0 0 s8     0000000000000001 s9     0000022200d00000
E/TC:0 0 s10    000000008e0b2fb0 s11    000000008e0d1350
E/TC:0 0 t3     000000008001902e t4     000000008004ddaf
E/TC:0 0 t5     0000000000000001 t6     fffffffffffffff6
E/TC:0 0 TEE load address @ 0x8e000000
E/TC:0 0 Call stack:
E/TC:0 0  0x8e0237a8 tee_mm_init at optee_os/core/mm/tee_mm.c:68
E/TC:0 0  0x8e0234a0 core_mmu_init_ta_ram at optee_os/core/mm/core_mmu.c:2525
E/TC:0 0  0x8e000950 init_tee_runtime at optee_os/core/arch/riscv/kernel/boot.c:61
E/TC:0 0  0x8e000a88 boot_init_primary_late at optee_os/core/arch/riscv/kernel/boot.c:120
E/TC:0 0  0x8e0000ec reset_primary at optee_os/core/arch/riscv/kernel/entry.S:181
E/TC:0 0 Panic 'unhandled page fault abort' at ?:0
E/TC:0 0 TEE load address @ 0x8e000000
E/TC:0 0 Call stack:
E/TC:0 0  0x8e0046cc read_pc at optee_os/core/arch/riscv/kernel/unwind_rv.c:146
E/TC:0 0  0x8e00478a print_kernel_stack at optee_os/core/arch/riscv/kernel/unwind_rv.c:51
E/TC:0 0  0x8e013ab4 __do_panic at optee_os/core/kernel/panic.c:26 (discriminator 32)
E/TC:0 0  0x8e0016b2 abort_handler at optee_os/core/arch/riscv/kernel/abort.c:384
E/TC:0 0  0x8e001f90 thread_abort_handler at optee_os/core/arch/riscv/kernel/thread_arch.c:223
E/TC:0 0  0x8e001fe2 thread_exception_handler at optee_os/core/arch/riscv/kernel/thread_arch.c:234
E/TC:0 0  0x8e0020b0 thread_trap_handler at optee_os/core/arch/riscv/kernel/thread_arch.c:277
E/TC:0 0  0x8e0002b8 trap_from_kernel at optee_os/core/arch/riscv/kernel/thread_rv.S:124
E/TC:0 0  0x8e0234a0 core_mmu_init_ta_ram at optee_os/core/mm/core_mmu.c:2525
E/TC:0 0  0x8e000950 init_tee_runtime at optee_os/core/arch/riscv/kernel/boot.c:61
E/TC:0 0  0x8e000a88 boot_init_primary_late at optee_os/core/arch/riscv/kernel/boot.c:120
E/TC:0 0  0x8e0000ec reset_primary at optee_os/core/arch/riscv/kernel/entry.S:181
```

**Comment by jforissier** [comment]
> Hi @jforissier
> 
> > Hmm, I guess the problem is stack pointer.
> > In RISC-V's entry.S, the stack pointer is not assigned as thread stack between boot_init_primary_early() and boot_init_primary_late().
> 
> After fixing the stack pointer between those two functions, I got correct unwinding.

Cool! :sunglasses: 

> I think this PR is good. I will create another PR to fix that after more checking.

I'd rather have the SP fix in this series unless you think it will delay merging this too much...
 
> ```
> cat dump.txt | ./optee_os/scripts/symbolize.py -d ./optee_os/out/riscv/core/
> E/TC:0 0 Core data-abort at address 0x0 (translation fault)
> E/TC:0 0 cpu    #0
> E/TC:0 0 cause  000000000000000f epc    000000008e0237a8
> E/TC:0 0 tval   0000000000000000 status 8000000200046100
> E/TC:0 0 ra     000000008e023750 sp     000000008e0d1290
> E/TC:0 0 gp     000000008e086268 tp     0000000000000018
> E/TC:0 0 t0     0000000000000001 t1     000000008e0d1400
> E/TC:0 0 t2     000000008e00016e s0     000000008e0d1390
> E/TC:0 0 s1     000000008e0d12b0 a0     000000008e0b2fb0
> E/TC:0 0 a1     0000000000000222 a2     0000000000000018
> E/TC:0 0 a3     0000000000000000 a4     0000000000000000
> E/TC:0 0 a5     0000000000000001 a5     0000000000000001
> E/TC:0 0 a6     0000000091c00000 a7     0000000091dfffff
> E/TC:0 0 s2     000000008e05b860 s3     000002228e0b2fb0
> E/TC:0 0 s4     000000008e086268 s5     000000008e0d1320
> E/TC:0 0 s6     000000008e05be76 s7     0000000000000018
> E/TC:0 0 s8     0000000000000001 s9     0000022200d00000
> E/TC:0 0 s10    000000008e0b2fb0 s11    000000008e0d1350
> E/TC:0 0 t3     000000008001902e t4     000000008004ddaf
> E/TC:0 0 t5     0000000000000001 t6     fffffffffffffff6
> E/TC:0 0 TEE load address @ 0x8e000000
> E/TC:0 0 Call stack:
> E/TC:0 0  0x8e0237a8 tee_mm_init at optee_os/core/mm/tee_mm.c:68
> E/TC:0 0  0x8e0234a0 core_mmu_init_ta_ram at optee_os/core/mm/core_mmu.c:2525
> E/TC:0 0  0x8e000950 init_tee_runtime at optee_os/core/arch/riscv/kernel/boot.c:61
> E/TC:0 0  0x8e000a88 boot_init_primary_late at optee_os/core/arch/riscv/kernel/boot.c:120
> E/TC:0 0  0x8e0000ec reset_primary at optee_os/core/arch/riscv/kernel/entry.S:181
> E/TC:0 0 Panic 'unhandled page fault abort' at ?:0
> ```

There's a small glitch here... why no file/line number? Is there a problem with the panic macros?

**Comment by gagachang** [comment]
> I'd rather have the SP fix in this series unless you think it will delay merging this too much...
Sure, I am preparing it.

> There's a small glitch here... why no file/line number? Is there a problem with the panic macros?
I did not enable `CFG_TEE_CORE_DEBUG`, only `CFG_UNWIND` is enabled.

**Comment by gagachang** [comment]
Hello @jforissier 

The new commit is added to solve that stack pointer issue.
Please check it. Thanks!

File: core/arch/riscv/kernel/unwind_rv.c [hunk 1]
```diff
@@ -0,0 +1,53 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*-
+ * Copyright (c) 2015 Linaro Limited
```
Comment (jforissier) [inline core/arch/riscv/kernel/unwind_rv.c:3]: Maybe add your own copyright here?

**Review by jforissier — CHANGES_REQUESTED** [review]
General comment, please remove `kernel:` from the commit subjects, it is somewhat redundant with `core:`.

- For commits:
"libunw: Implement RISC-V stack unwinding"
"core: riscv: Implement API to read PC and FP" -- preferably rewrite as "core: riscv: add read_fp() and read_pc()"

`Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`

- For commits:
"core: riscv: kernel: Save and restore register s0 when trap occurs"
"core: riscv: Add core compiler flags for stack unwinding"
"core: riscv: kernel: Update the kernel mode unwinding"
"core: riscv: kernel: Implement function to print kernel stack" -- note one comment below
`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

As for "core: riscv: kernel: Correct sp before boot_init_primary_late()" I'd rather have it first so that everything works when the actual unwinding code is introduced. Remove `kernel:` then you can add:
`Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`

File: core/arch/riscv/kernel/unwind_rv.c (hunk 1 above)
Comment (gagachang) [reply core/arch/riscv/kernel/unwind_rv.c:3]: OK I am adding it.

Comment (gagachang) [reply core/arch/riscv/kernel/unwind_rv.c:4]: Done

File: lib/libunw/unwind_riscv.c [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+#include <util.h>
+
+void __weak ftrace_map_lr(uint64_t *lr __unused)
+{}
+
+bool unwind_stack_riscv(struct unwind_state_riscv *frame,
+			vaddr_t stack, size_t stack_size)
+{
+	vaddr_t fp = frame->fp;
+	struct unwind_state_riscv *caller_state;
```
Comment (etienne-lms) [inline lib/libunw/unwind_riscv.c:47]: = NULL

File: lib/libunw/unwind_riscv.c [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+
+#include <compiler.h>
+#include <string.h>
+#include <trace.h>
+#include <types_ext.h>
+#include <unw/unwind.h>
+#include <util.h>
+
+void __weak ftrace_map_lr(uint64_t *lr __unused)
+{}
```
Comment (etienne-lms) [inline lib/libunw/unwind_riscv.c:41]: line break between braces:
```
{
}

File: lib/libunw/unwind_riscv.c [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+/*-
+ * Copyright (c) 2023 Andes Technology Corporation
+ * Copyright (c) 2015-2019 Linaro Limited
+ * Copyright (c) 2015 The FreeBSD Foundation
+ * All rights reserved.
+ *
+ * This software was developed by Semihalf under
+ * the sponsorship of the FreeBSD Foundation.
+ *
+ * Redistribution and use in source and binary forms, with or without
```
Comment (etienne-lms) [inline lib/libunw/unwind_riscv.c:11]: the text starting from here could be omitted as it is standar dBSD-2-Clause terms.

File: core/arch/riscv/include/riscv.h [hunk 5]
```diff
@@ -133,6 +133,24 @@ static inline __noprof unsigned long read_tp(void)
 	return tp;
 }
 
+static inline __noprof unsigned long read_fp(void)
+{
+	unsigned long fp;
```
Comment (etienne-lms) [inline core/arch/riscv/include/riscv.h:138]: = 0

File: core/arch/riscv/include/riscv.h [hunk 6]
```diff
@@ -133,6 +133,24 @@ static inline __noprof unsigned long read_tp(void)
...
+
+	asm volatile ("mv %0, fp" : "=r" (fp));
+
+	return fp;
+}
+
+static inline __noprof unsigned long read_pc(void)
+{
+	unsigned long pc;
```
Comment (etienne-lms) [inline core/arch/riscv/include/riscv.h:147]: = 0

File: core/arch/riscv/kernel/unwind_rv.c [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+/*-
+ * Copyright (c) 2023 Andes Technology Corporation
+ * Copyright (c) 2015 Linaro Limited
+ * Copyright (c) 2015 The FreeBSD Foundation
+ * All rights reserved.
+ *
+ * This software was developed by Semihalf under
+ * the sponsorship of the FreeBSD Foundation.
+ *
+ * Redistribution and use in source and binary forms, with or without
```
Comment (etienne-lms) [inline core/arch/riscv/kernel/unwind_rv.c:11]: could remove the full license terms text as standard BSD 2 Clause terms implicitly mention through the SPDX licence tag.

**Review by etienne-lms — COMMENTED** [review]
minor comments

File: lib/libunw/unwind_riscv.c (hunk 2 above)
Comment (gagachang) [reply lib/libunw/unwind_riscv.c]: Done

File: lib/libunw/unwind_riscv.c (hunk 3 above)
Comment (gagachang) [reply lib/libunw/unwind_riscv.c]: Done

File: core/arch/riscv/kernel/unwind_rv.c (hunk 7 above)
Comment (gagachang) [reply core/arch/riscv/kernel/unwind_rv.c]: Done

File: core/arch/riscv/include/riscv.h (hunk 6 above)
Comment (gagachang) [reply core/arch/riscv/include/riscv.h]: Done

**Review by jforissier — APPROVED** [review]

**Review by etienne-lms — APPROVED** [review]
`Acked-by: Etienne Carriere <etienne.carriere@foss.st.com>`

**Comment by gagachang** [comment]
> 

Hi @etienne-lms 

Is this tag applicable to all commits?

**Comment by etienne-lms** [comment]
Yes it is. Sorry, I implicitly refer to all commits in a P-R series when I don't refer to specific commit header lines.

**Pull request merged**